### Homework 3 Numpy Arrays and Pandas

1. Open `cars.csv` in jupyter and look at the data. (Double click `cars.csv` to open in a new tab).
2. We are converting each **column** into a numpy array.
3. Notice we have text, integer and float data. 
4. Take care not to overwrite your starter code.

In [17]:
#Starter code. Do NOT modify this cell
import pandas as pd
import numpy as np

# Read in dataset
df = pd.read_csv('cars.csv',delimiter=';')

# Convert each column to an np.array
car = df.Car.to_numpy()
mpg = df.MPG.to_numpy()
cylinders = df.Cylinders.to_numpy()
displacement = df.Displacement.to_numpy()
horsepower = df.Horsepower.to_numpy()
weight = df.Weight.to_numpy()
acceleration = df.Acceleration.to_numpy()
model = df.Model.to_numpy()
origin = df.Origin.to_numpy()

# create an array of headers
headers = np.array(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight','Acceleration', 'Model', 'Origin'])


1. (2 pts) What are the numpy datatypes for the `car`,`mpg`, `horsepower` and `origin` arrays? (Hint: Not `type(car)` and use `print()` to print the numpy type.)

In [18]:
# Using dtype attribute to find the datatypes for each array.
print("Car:", car.dtype)
print("Mpg:", mpg.dtype)
print("Horsepower:", horsepower.dtype)
print("Origin:", origin.dtype)

car: object
mpg: float64
horsepower: float64
origin: object


***The numpy datatype for car and origin are objects while the data type for mpg and horsepower are float64.***

2. (1 pt) `numpy` is often imported as `np`. What is this action called? Why might you do this? 

***The action of importing numpy as np is called aliasing. The reason you might do this is to make your code easier to type out. Instead of typing out numpy every time you want to use a function from numpy, you could just use np.***

3. (2 pts) Each of the arrays (cars, mpg, cylinders, etc..) has how many dimensions? Show how to find this using code. 

In [19]:
# Using the .ndim attribute to figure out how many dimensions there are for each array.
print("Car:", car.ndim)
print("MPG:", mpg.ndim)
print("Cylinders:", cylinders.ndim)
print("Displacement:", displacement.ndim)
print("Horsepower:", horsepower.ndim)
print("Weight:", weight.ndim)
print("Acceleration:", acceleration.ndim)
print("Model:", model.ndim)
print("Origin:", origin.ndim)

car: 1
mpg: 1
cylinders: 1
displacement: 1
horsepower: 1
weight: 1
acceleration: 1
model: 1
origin: 1


***It looks like there is only one dimension for each of the arrays.***

4. (3 pts) What is the ___ of `horsepower`?
- sum
- min
- max
- median
- Optional: mode (May need to google help) (1 pt extra credit)

In [20]:
# The basic functions for finding the respective ___ for the question above.
print("Sum:",horsepower.sum())
print("Min:",horsepower.min())
print("Max:",horsepower.max())
print("Median:",np.median(horsepower))

# Code to find mode from "in plain english" website, because I was unsure if I could import statistics from scipy.
def find_mode(np_array):
    vals,counts = np.unique(np_array, return_counts=True)
    index = np.argmax(counts)
    return(vals[index])

list = np.array([horsepower])

print("Mode:",find_mode(list))

Sum: 42033.0
Min: 0.0
Max: 230.0
Median: 93.5
Mode: 150.0


5. (2 pts) Select the last element in each subarray (do this with one operation.)

In [21]:
select_more = np.array([[1,2,3],
                      [4,5,6],
                      [7,8,9]])
# returns array([3, 6, 9])

In [22]:
# references were very helpful :)
# : is used to broadcast because I needed to repeat for each row.
# -1 just means len - 1 which is just the last number in the row.
select_more[: , -1]

array([3, 6, 9])

6. (2 pts) Use the built-in `np.lookfor()` function to search for `'log'`.

Use the function you find and apply it to `mpg`.

***The np.lookfor() function was removed in numpy version 2.1; therefore, I am unable to use it to search for log or apply it to MPG.***

7. (2 pts) Use boolean indexing to create an array of `True/False` using `origin` and the value 'US'.

In [23]:
# Make code that checks if origin is == to "US".
boolean_indexing = (origin == "US")
# Print to see array of True/False.
print(boolean_indexing)

[ True  True  True  True  True  True  True  True  True  True False  True
  True  True  True  True  True  True  True  True False  True  True  True
 False False False False False False  True  True  True  True  True False
  True False  True False  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True False False False
 False False False  True False  True False  True  True  True  True  True
  True  True  True  True  True  True False  True  True  True  True False
 False False False  True False False  True False  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True False  True  True  True  True  True False  True False False  True
  True False  True  True False False False False  True False False  True
  True  True  True  True False  True False  True  True  True  True  True
  True  True  True  True False False False False False  True False False
 False False False  True  True  True  True  True  T

8. (2 pts) Apply your index (from above) to the `car` array. You should now only see cars that are of US origin. 

In [24]:
# Create New Variable and apply boolean_indexing to car array.
american_cars = car[boolean_indexing]
# Print to see result.
print(american_cars)

['Chevrolet Chevelle Malibu' 'Buick Skylark 320' 'Plymouth Satellite'
 'AMC Rebel SST' 'Ford Torino' 'Ford Galaxie 500' 'Chevrolet Impala'
 'Plymouth Fury iii' 'Pontiac Catalina' 'AMC Ambassador DPL'
 'Chevrolet Chevelle Concours (sw)' 'Ford Torino (sw)'
 'Plymouth Satellite (sw)' 'AMC Rebel SST (sw)' 'Dodge Challenger SE'
 "Plymouth 'Cuda 340" 'Ford Mustang Boss 302' 'Chevrolet Monte Carlo'
 'Buick Estate Wagon (sw)' 'Plymouth Duster' 'AMC Hornet' 'Ford Maverick'
 'AMC Gremlin' 'Ford F250' 'Chevy C20' 'Dodge D200' 'Hi 1200D'
 'Chevrolet Vega 2300' 'Ford Pinto' 'AMC Gremlin'
 'Plymouth Satellite Custom' 'Chevrolet Chevelle Malibu' 'Ford Torino 500'
 'AMC Matador' 'Chevrolet Impala' 'Pontiac Catalina Brougham'
 'Ford Galaxie 500' 'Plymouth Fury iii' 'Dodge Monaco (sw)'
 'Ford Country Squire (sw)' 'Pontiac Safari (sw)'
 'AMC Hornet Sportabout (sw)' 'Chevrolet Vega (sw)' 'Pontiac Firebird'
 'Ford Mustang' 'Mercury Capri 2000' 'Plymouth Cricket'
 'Dodge Colt Hardtop' 'Chevrolet Vega' 'Ford

### Use numpy or pandas to answer the following questions.

9. (3 pts) What is the average horsepower for `Europe` cars? (Hint: create a boolean index of 'US' apply it to horsepower and perform the calculation.)

In [29]:

# Re-use code from above, swap "US" with "Europe".
boolean_indexing_1 = (origin == "Europe")
# Re-use code from above with mean() function afterwards.
european_cars = horsepower[boolean_indexing_1].mean()
# Print to see result.
print("Average Horsepower:",european_cars)

Average Horsepower: 78.78082191780823


***The average horsepower of European cars is 78.78 horsepower***

10. (3 pts) On average are Japanese cars more fuel-efficient than US cars? Show the code that you used to determine this.  

In [26]:
cars_path = "cars.csv"

cars_df = pd.read_csv("cars.csv", delimiter =';')

grouped_df = cars_df.groupby('Origin')

mean_df = grouped_df[['MPG']].mean()

print(mean_df)

              MPG
Origin           
Europe  26.745205
Japan   30.450633
US      19.688189


***On average, Japanese cars are more fuel-efficient than US cars.***

11. (3 pts) What is the correlation coefficient of `mpg` and `weight`? What is a correlation coefficient? What does the result indicate about the relationship between mpg and weight? 

In [27]:
print(np.corrcoef(cars_df["MPG"], cars_df["Weight"]))

[[ 1.         -0.78750994]
 [-0.78750994  1.        ]]


***The correlation coefficienct is the measure of linear correlation within a dataset. The result from the code indicates there is a fairly strong negative relationship between MPG and Weight***

Extra credit: (3 pts) Create a new pandas dataframe that has the average mpg, weight, and horsepower per country of origin

In [28]:
cars_path = "cars.csv"

cars_df = pd.read_csv("cars.csv", delimiter =';')

grouped_df = cars_df.groupby('Origin')

mean_df = grouped_df[['MPG', 'Weight', 'Horsepower']].mean()

print(mean_df)

              MPG       Weight  Horsepower
Origin                                    
Europe  26.745205  2431.493151   78.780822
Japan   30.450633  2221.227848   79.835443
US      19.688189  3372.700787  118.011811
